clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1933,11 +1941,81 @@ private Map<String, String> topicPartitionTags(TopicPartition tp) {
+            if (fetchTarget == null || client.isUnavailable(fetchTarget)) {
+                log.debug("Skip sending close session request to broker {} since it is not reachable", fetchTarget);
+                continue;
+            }
+
+            log.debug("Sending close request for fetch session: {} to node: {}", sessionId, fetchTarget);
+            final RequestFuture<ClientResponse> responseFuture = sendFetchRequestToNode(sessionHandler.newBuilder().build(), fetchTarget);
+            responseFuture.addListener(new RequestFutureListener<ClientResponse>() {
+                @Override
+                public void onSuccess(ClientResponse value) {
+                    log.info("Successfully sent a close message for fetch session: {} to node: {}", sessionId, fetchTarget);
+                }
+
+                @Override
+                public void onFailure(RuntimeException e) {
+                    log.info("Unable to a close message for fetch session: {} to node: {}. " +
+                        "This may result in unnecessary fetch sessions at the broker.", sessionId, fetchTarget, e);
+                }
+            });
+
+            requestFutures.add(responseFuture);
+        }
+
+        // Poll to ensure that request has been written to the socket. Wait until either the timer has expired or until
+        // all requests have received a response.
+        do {
+            client.poll(timer, null, true);
+        } while (timer.notExpired() && !requestFutures.stream().allMatch(RequestFuture::isDone));
+
+        if (!requestFutures.stream().allMatch(RequestFuture::isDone)) {
+            // we ran out of time before completing all futures. It is ok since we don't want to block the shutdown
+            // here.
+            log.warn("All requests couldn't be sent in the specific timeout period {}ms. " +
+                "This may result in unnecessary fetch sessions at the broker. Consider increasing the timeout passed for " +
+                "KafkaConsumer.close(Duration timeout)", timer.timeoutMs());
+        }
+    }
+
+    public void close(final Timer timer) {
+        if (isClosed.get()) {
+            log.info("Fetcher {} is already closed.", this);
+            return;
+        }
+
+        // Shared states (e.g. sessionHandlers) could be accessed by multiple threads (such as heartbeat thread), hence,
+        // it is necessary to acquire a lock on the fetcher instance before modifying the states.
+        synchronized (Fetcher.this) {
+            // we do not need to re-enable wakeups since we are closing already
+            client.disableWakeups();
+            if (nextInLineFetch != null)
+                nextInLineFetch.drain();
+            maybeCloseFetchSessions(timer);
+            Utils.closeQuietly(decompressionBufferSupplier, "decompressionBufferSupplier");
+            sessionHandlers.clear();
+        }
+        this.isClosed.compareAndSet(false, true);
```

Comment:
I think we can set the `isClosed` earlier, so that it will not enter the synchronize block multiple times for different threads. ex:

```java
if (this.isClosed.compareAndSet(false, true)) {
          // Shared states (e.g. sessionHandlers) could be accessed by multiple threads (such as heartbeat thread), hence,
        // it is necessary to acquire a lock on the fetcher instance before modifying the states.
        synchronized (Fetcher.this) {
            // we do not need to re-enable wakeups since we are closing already
            client.disableWakeups();
            if (nextInLineFetch != null)
                nextInLineFetch.drain();
            maybeCloseFetchSessions(timer);
            Utils.closeQuietly(decompressionBufferSupplier, "decompressionBufferSupplier");
            sessionHandlers.clear();
        }
} else {
  log.info("Fetcher {} is already closed.", this);
}
```
WDYT?

Reply:
Good idea. Changed in latest commit.